### ISSUE TITLE
CUDA out Of Error while Loading From Checkpoint

### ISSUE THREAD

**Issue by chinge55** [Lightning-AI/lightning#18991]
### Bug description

I have a pytorch lightning model that I have trained and the checkpoints are saved successfully. 
```python
model = LitAutoEncoder(*params).to(torch.device("cuda:1") 
```
The above code doesn't show any error and the model looks just fine. 

On contrast, if I run the code: 
```python
model = LitAutoEncoder(*params).to(torch.device("cuda:1") .load_from_checkpoint(ckpt_path)
```
It gives me a Cuda out of Memory error: 

This should not be happening because the model only takes around 500Mb of memory in GPU. I have tried on multiple GPUs of memories ranging from (10GB, 40GB and 80GB) of VRAM respectively. Still the error persists.

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
model.py

import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np


def get_conv_block(filter_in, filter_out, kernel_size):
    return nn.Sequential(
        nn.Conv2d(
            in_channels=filter_in,
            out_channels=filter_out,
            kernel_size=kernel_size,
            stride=1,
        ),
        nn.ReLU(),
        nn.MaxPool2d(kernel_size=2, stride=2),
    )


class Encoder(nn.Module):
    def __init__(self, input_size=(3, 200, 200), latent_dim=128):
        super().__init__()
        self.input_size = input_size
        self.conv_block_1 = get_conv_block(3, 32, 3)
        self.conv_block_2 = get_conv_block(32, 32, 3)
        self.conv_block_3 = get_conv_block(32, 32, 3)

    def forward(self, x):
        x = self.conv_block_1(x)
        x = self.conv_block_2(x)
        x = self.conv_block_3(x)
        return x


def get_deconv_block(in_channels=32, out_channels=32, stride=1):
    return nn.Sequential(
        nn.Upsample(scale_factor=2, mode="nearest"),
        nn.ConvTranspose2d(
            in_channels=in_channels,
            out_channels=out_channels,
            kernel_size=3,
            stride=stride,
        ),
        nn.ReLU(),
    )


class Decoder(nn.Module):
    def __init__(self, input_size=(32, 23, 23)):
        super(Decoder, self).__init__()
        self.deconv_block_1 = get_deconv_block()
        self.deconv_block_2 = get_deconv_block(stride=1)
        self.deconv_block_3 = get_deconv_block(out_channels=12)
        self.out_deconv_block = nn.ConvTranspose2d(
            in_channels=12, out_channels=3, kernel_size=3, stride=1
        )

    def forward(self, x):
        x = self.deconv_block_1(x)
        x = self.deconv_block_2(x)
        x = self.deconv_block_3(x)
        x = self.out_deconv_block(x)
        x = torch.sigmoid(x)  # For valid image pixel values (0-1)
        return x


class AutoEncoder(nn.Module):
    def __init__(self, input_size=(3, 200, 200), hidden_size=1024):
        super(AutoEncoder, self).__init__()
        self.encoder = Encoder(input_size, hidden_size)
        self.decoder = Decoder()

    def forward(self, x):
        x = self.encoder(x)
        x = self.decoder(x)
        return x

model_lightning.py 
```python
from .model import AutoEncoder
import pytorch_lightning as pl
import torch.nn.functional as F
import torch


class LitAutoEncoder(pl.LightningModule):
    def __init__(self, min_value, max_value):
        super().__init__()
        self.model = AutoEncoder()
        self.min_value = min_value
        self.max_value = max_value

    def training_step(self, batch, batch_idx):
        x, y = batch
        x = self.model(x)
        loss = F.mse_loss(x, y)
        self.log(
            "train_losss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True
        )
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        x = self.model(x)
        loss = F.mse_loss(x, y)
        ######################LOGGING#############################
        self.log("val_loss", loss, on_epoch=True)
        mae = F.l1_loss(x, y)
        self.log("val_mae", mae, on_epoch=True)
        max_ae = torch.max(torch.abs(x - y))
        self.log("val_max_ae", max_ae, on_epoch=True)

        # if batch_idx % 100 == 0:
        #     self.logger.experiment.add_images(
        #         "val_images", x, global_step=self.global_step
        #     )

    def test_step(self, batch, batch_idx):
        x, y = batch
        x = self.model(x)
        loss = F.mse_loss(x, y)
        self.log("test_loss", loss, on_epoch=True)
        # De normalize x and y before calculating MAE
        x = x * (self.max_value - self.min_value) + self.min_value
        y = y * (self.max_value - self.min_value) + self.min_value
        mae = F.l1_loss(x, y)
        self.log("test mae (Denormalized)", mae, on_epoch=True)
        max_ae = torch.max(torch.abs(x - y))
        self.log("test max_ae (Denormalized)", max_ae, on_epoch=True)
        # get all 3 channels from the image
        # TODO: Proper Testing individual channels and logging them.

        self.logger.experiment.add_images(
            "test_images", y, global_step=self.global_step
        )
        self.logger.experiment.add_images(
            "test_images_predicted", x, global_step=self.global_step
        )

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer
```
```


### Error messages and logs
```
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 2080 Ti
                - NVIDIA GeForce RTX 2080 Ti
                - NVIDIA GeForce RTX 2080 Ti
                - NVIDIA GeForce RTX 2080 Ti
                - NVIDIA GeForce RTX 2080 Ti
                - NVIDIA GeForce RTX 2080 Ti
                - NVIDIA RTX A6000
                - NVIDIA A100 80GB PCIe
        - available:         True
        - version:           11.7
* Lightning:
        - lightning-utilities: 0.9.0
        - pytorch-lightning: 2.0.6
        - torch:             2.0.1
        - torchaudio:        2.0.2
        - torchmetrics:      1.0.3
        - torchvision:       0.15.2
* Packages:
        - absl-py:           1.4.0
        - aiofiles:          22.1.0
        - aiohttp:           3.8.1
        - aiosignal:         1.3.1
        - aiosqlite:         0.18.0
        - anyio:             3.5.0
        - argon2-cffi:       21.3.0
        - argon2-cffi-bindings: 21.2.0
        - asttokens:         2.0.5
        - async-timeout:     4.0.2
        - attrs:             22.1.0
        - babel:             2.11.0
        - backcall:          0.2.0
        - beautifulsoup4:    4.12.2
        - bleach:            4.1.0
        - blinker:           1.6.2
        - bottleneck:        1.3.5
        - brotlipy:          0.7.0
        - cachetools:        5.3.1
        - certifi:           2023.7.22
        - cffi:              1.15.1
        - charset-normalizer: 2.0.4
        - click:             8.1.6
        - comm:              0.1.2
        - contourpy:         1.0.5
        - cryptography:      39.0.1
        - cycler:            0.11.0
        - debugpy:           1.5.1
        - decorator:         5.1.1
        - defusedxml:        0.7.1
        - entrypoints:       0.4
        - exceptiongroup:    1.1.2
        - executing:         0.8.3
        - fastjsonschema:    2.16.2
        - filelock:          3.9.0
        - fonttools:         4.25.0
        - frozenlist:        1.3.3
        - fsspec:            2023.6.0
        - gmpy2:             2.1.2
        - google-auth:       2.22.0
        - google-auth-oauthlib: 1.0.0
        - grpcio:            1.48.2
        - idna:              3.4
        - importlib-metadata: 6.0.0
        - importlib-resources: 5.2.0
        - iniconfig:         2.0.0
        - ipykernel:         6.19.2
        - ipython:           8.12.0
        - ipython-genutils:  0.2.0
        - ipywidgets:        8.0.4
        - jedi:              0.18.1
        - jinja2:            3.1.2
        - json5:             0.9.6
        - jsonschema:        4.17.3
        - jupyter:           1.0.0
        - jupyter-client:    7.0.6
        - jupyter-console:   6.6.3
        - jupyter-core:      5.3.0
        - jupyter-events:    0.6.3
        - jupyter-server:    1.23.4
        - jupyter-server-fileid: 0.9.0
        - jupyter-server-terminals: 0.4.4
        - jupyter-server-ydoc: 0.8.0
        - jupyter-ydoc:      0.2.4
        - jupyterlab:        3.6.3
        - jupyterlab-pygments: 0.1.2
        - jupyterlab-server: 2.22.0
        - jupyterlab-widgets: 3.0.5
        - kiwisolver:        1.4.4
        - lightning-utilities: 0.9.0
        - lxml:              4.9.1
        - markdown:          3.4.4
        - markupsafe:        2.1.1
        - matplotlib:        3.7.1
        - matplotlib-inline: 0.1.6
        - mistune:           0.8.4
        - mkl-fft:           1.3.6
        - mkl-random:        1.2.2
        - mkl-service:       2.4.0
        - mpmath:            1.2.1
        - multidict:         6.0.2
        - munkres:           1.1.4
        - nbclassic:         0.5.5
        - nbclient:          0.5.13
        - nbconvert:         6.5.4
        - nbformat:          5.7.0
        - nest-asyncio:      1.5.6
        - networkx:          2.8.4
        - notebook:          6.5.4
        - notebook-shim:     0.2.2
        - numexpr:           2.8.4
        - numpy:             1.24.3
        - oauthlib:          3.2.2
        - packaging:         23.0
        - pandas:            1.5.3
        - pandocfilters:     1.5.0
        - parso:             0.8.3
        - pexpect:           4.8.0
        - pickleshare:       0.7.5
        - pillow:            9.4.0
        - pip:               23.1.2
        - pkgutil-resolve-name: 1.3.10
        - platformdirs:      2.5.2
        - pluggy:            1.2.0
        - ply:               3.11
        - prometheus-client: 0.14.1
        - prompt-toolkit:    3.0.36
        - protobuf:          3.20.3
        - psutil:            5.9.0
        - ptyprocess:        0.7.0
        - pure-eval:         0.2.2
        - pyasn1:            0.4.8
        - pyasn1-modules:    0.2.7
        - pycparser:         2.21
        - pygments:          2.15.1
        - pyjwt:             2.8.0
        - pyopenssl:         23.0.0
        - pyparsing:         3.0.9
        - pyqt5-sip:         12.11.0
        - pyrsistent:        0.18.0
        - pysocks:           1.7.1
        - pytest:            7.4.0
        - python-dateutil:   2.8.2
        - python-json-logger: 2.0.7
        - pytorch-lightning: 2.0.6
        - pytz:              2022.7
        - pyu2f:             0.1.5
        - pyyaml:            6.0
        - pyzmq:             19.0.2
        - qtconsole:         5.4.2
        - qtpy:              2.2.0
        - requests:          2.29.0
        - requests-oauthlib: 1.3.1
        - rfc3339-validator: 0.1.4
        - rfc3986-validator: 0.1.1
        - rsa:               4.9
        - send2trash:        1.8.0
        - setuptools:        67.8.0
        - sip:               6.6.2
        - six:               1.16.0
        - sniffio:           1.2.0
        - soupsieve:         2.4
        - stack-data:        0.2.0
        - sympy:             1.11.1
        - tensorboard:       2.13.0
        - tensorboard-data-server: 0.7.0
        - terminado:         0.17.1
        - tinycss2:          1.2.1
        - toml:              0.10.2
        - tomli:             2.0.1
        - torch:             2.0.1
        - torchaudio:        2.0.2
        - torchmetrics:      1.0.3
        - torchvision:       0.15.2
        - tornado:           6.2
        - tqdm:              4.66.1
        - traitlets:         5.7.1
        - triton:            2.0.0
        - typing-extensions: 4.6.3
        - urllib3:           1.26.16
        - wcwidth:           0.2.5
        - webencodings:      0.5.1
        - websocket-client:  0.58.0
        - werkzeug:          2.3.6
        - wheel:             0.38.4
        - widgetsnbextension: 4.0.5
        - y-py:              0.5.9
        - yarl:              1.7.2
        - ypy-websocket:     0.8.2
        - zipp:              3.11.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.17
        - release:           4.15.0-197-generic
        - version:           #208-Ubuntu SMP Tue Nov 1 17:23:37 UTC 2022

</details>

### More info

If this information is not enough to recreate the bug then I can also create a google colab and upload a link here.  
Pytorch Lightning Version: '2.0.6'
Operating System: Ubuntu
Running On: Anaconda, Python 3.8.17
